Sources/Subprocess/IO/AsyncIO+Linux.swift

```diff
@@ -0,0 +1,507 @@
+
+import _SubprocessCShims
+import Synchronization
+
+private typealias SignalStream = AsyncThrowingStream<Bool, any Error>
+private let _epollEventSize = 256
+private let _registration: Mutex<
+    [PlatformFileDescriptor : SignalStream.Continuation]
+> = Mutex([:])
+
+final class AsyncIO: Sendable {
+
+    typealias OutputStream = AsyncThrowingStream<AsyncBufferSequence.Buffer, any Error>
+
+    private final class MonitorThreadContext {
+        let epollFileDescriptor: CInt
+        let shutdownFileDescriptor: CInt
+
+        init(
+            epollFileDescriptor: CInt,
+            shutdownFileDescriptor: CInt
+        ) {
+            self.epollFileDescriptor = epollFileDescriptor
+            self.shutdownFileDescriptor = shutdownFileDescriptor
+        }
+    }
+
+    private enum Event {
+        case read
+        case write
+    }
+
+    private struct State {
+        let epollFileDescriptor: CInt
+        let shutdownFileDescriptor: CInt
+        let monitorThread: pthread_t
+    }
+
+    static let shared: AsyncIO = AsyncIO()
+
+    private let state: Result<State, SubprocessError>
+
+    private init() {
+        // Create main epoll fd
+        let epollFileDescriptor = epoll_create1(CInt(EPOLL_CLOEXEC))
+        guard epollFileDescriptor >= 0 else {
+            let error = SubprocessError(
+                code: .init(.asyncIOFailed("epoll_create1 failed")),
+                underlyingError: .init(rawValue: errno)
+            )
+            self.state = .failure(error)
+            return
+        }
+        // Create shutdownFileDescriptor
+        let shutdownFileDescriptor = eventfd(0, CInt(EFD_NONBLOCK | EFD_CLOEXEC))
+        guard shutdownFileDescriptor >= 0 else {
+            let error = SubprocessError(
+                code: .init(.asyncIOFailed("eventfd failed")),
+                underlyingError: .init(rawValue: errno)
+            )
+            self.state = .failure(error)
+            return
+        }
+
+        // Register shutdownFileDescriptor with epoll
+        var event = epoll_event(
+            events: EPOLLIN.rawValue,
+            data: epoll_data(fd: shutdownFileDescriptor)
+        )
+        var rc = epoll_ctl(
+            epollFileDescriptor,
+            EPOLL_CTL_ADD,
+            shutdownFileDescriptor,
+            &event
+        )
+        guard rc == 0 else {
+            let error = SubprocessError(
+                code: .init(.asyncIOFailed(
+                    "failed to add shutdown fd \(shutdownFileDescriptor) to epoll list")
+                ),
+                underlyingError: .init(rawValue: errno)
+            )
+            self.state = .failure(error)
+            return
+        }
+
+        // Create thread data
+        let context = MonitorThreadContext(
+            epollFileDescriptor: epollFileDescriptor,
+            shutdownFileDescriptor: shutdownFileDescriptor
+        )
+        let threadContext = Unmanaged.passRetained(context)
+#if os(FreeBSD) || os(OpenBSD)
```

Comment:
No need for FreeBSD/OpenBSD checks here since they'd use a kqueue implementation (and a POSIX poll implementation is probably different enough that we'd want its own file rather than ifdef'ing in the Linux one)